Voxel pick api

jjhembd opened the pull request:
<!--
Thanks for the Pull Request!

Please review [Contribution Guide](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md) before opening your first Pull Request.

To ensure your Pull Request is reviewed and accepted quickly, please refer to our [Pull Request Guidelines](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md#pull-request-guidelines).

-->

# Description

This PR adds the public method `Scene.pickVoxel` to report information about a voxel cell under the cursor.

The return value from `Scene.pickVoxel` is an instance of the new `VoxelCell` class. This class includes getters to retrieve the values of the different properties contained in the voxel metadata.

## Testing plan

- Verify the additional specs in `PickingSpec` and `VoxelCellSpec` run successfully.
- Test mouseover picking in [this local Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=3Vp9b9s2E/8qnLc/lNaV8zZgiJNgqZuuBuImqLOsbRy0tERbXClKD0k5dgp/9+dIinp1mvV5NjRdkEASeXc8/o53ujslSLhUaEHJLRHoCHFyiwZE0iz2r8yYN+kE5nmQcIUpJ2LS6aLPE47QFEtyhldEHDiWYYznRKzMoD8TSXwhkgUNiTiRKx54mgk52kvKyAinYyIWNCA5p6M3zL8LlrMUTNOMsnCUhBkjZrZzIiVRsndJlioTRPZeY7hidoqFiobDSWfLCjCXrW5N6wsafNK6zzCTxEzNSRIkYX1M0Zgw2HZ1DHMaY0UTXg6ut/oTPuGBQfMzkgHhpIsCHBOB0RqAtQgbInvrg8qEh16+Mzt4lSwJG3KZkkAlYkSXlGvBRpwfkmk2H0fJ7UsBcuUFEWMCC4YgXYmMGNmzjAdaM9iLGlH+PMl4KD0Z4ZRsWavRGbLP6OjoqLC1XnesRy9XKfFPz86GF+Pz4YucByG7sdhJRAXnAJCGO8z3/IAlnDQt1pC8Qa3cQn17LZbw72CRbX87HxcEDMvLaTO81lsu5mCYzMBUIcytW2Dg5d8NhpP4f4HRUKsBhps1YOxsm58WJI7oqyABTwtIqH1lTPmcEe2ORj3ngjWcVESlnwOFzLVfDFcPxQbzVggrgG3YekkYgs9xCVrKekQq4fV+6SL9W2Hi2ieA/hoCVsISCFM35aQCuM1kLmlEFA6xwsYMV6eD/QptkMQp2I6ry3uYBlUC/+XZ+cnl3u6NQ9geDhkIrIJooDVpbEIPeTZcPGgEPxWJSrT2viD/yYhUL0ABEFjY0RyxJNX3csJzc7kwpEDeGVkQ1jW3b+3lnb28N3EpZ+3bY6P9oWBCx3DmikO/4UQVp82uVzNbw5D9kkzvxhEYu1xv31SmBeZhEo8JgbhmV7bKPqnI91f1xyV6ane2efhtRbw1oLWKGRFZoN6YNTX6Gl6vFNE12nYrSm31aw4Ghoqp1NaRCVsQ77qQf6Mp1/17nG6UMUW/M5/b7yL9+7h8rhDxgekzO4DdKeDfd+74EOTf3MOayv8bHU6j/g/6mxa0yd3+p/Us/O74YRX5kGFaAa+zeKqTLhJ6dQXdXhf6ZLkjWMPlS2DeVSQEEea8lLHJdcCNC28wdJ7ew9Y9MU678kuWYLW3eyIEXnkVHZ/UlnNIz4DPY0Qhk4D14XJY1xbdPX1anNKCemWpV3XqFYxUqCv0S0u/rNMvYaRG7/bEiEiJeAtMS9RDXg3cZ/Ca6m9meAcMqwbD6ksM74HhrsFw5xhKll7PMkVAHmUEzr1T8AkkrT8Dg1tfP+8V8+/N825r8ahibH3mOFQH1UPntfXVTr/jbzeWaotmJo/+uTUe5CekmpeYuuuVZF7UReAcENVghW4tm6njYEVRiFFLFz3sz92D8WPD0LKlpPWz3a9HB7M0wpYGmA3XjCWQeTmJTzUo9c0UfnNtN1T3D81xY6KYBkrkAfkvcu6UnHNBCP8a3t2Sd8oy8jWse5q1AkXBu7Y363bJUMhslQyBIFiRC0FjqijE3jR/hUKdm0kF0TDC8JT7ri1ZU0csAbA4WZATxjwHeiVmFkJ1qtPkxGGYF1LV7kSNzSvihVPqoLjruqmqlge1p4IkhvKfwWERdHlQfcgJ1raPkOtf1xwOJYa3khrjOGWQzZdVeYvyT6qU6baY7oGhMFS2L7Co9QBMY2CkNfFbSNUH7AvAdBzgqK8uE5PpgSbjNCKCeA0lprXZrou4OlWyrQ0oNM22141XcGvVStlTBXVT7VOZtiabCTyHEKDsmG7jHKCPi4SGxcwIU+69zB+GPM0Umklz7ULcSeAxuIs/OFsRQTGDktjeGHMVR8ONQpY0m2VSI5hL8uP8Bevnzj2f9ltcxo3u58E2P/vYbAZVIfkjooo8ckgWJNjzIOpv3Y/ATt6CaG+24jK1kA3P++bdMg4ws22RdstEz58yojcl673CU8ZoKgEC/4/fxr/s66qGxln8Boc0k91/hNQ4+oTbzY0heAZYKkZ0MLpMEjbFYkR45l1X4FTGVJNOIRdezWX5gXT6OenkKgBahIGDH1TqjErWY9F0ESw/MQ+3bMo3wwOtLEdYGtmtWEaXZrgvhbciK2o7fnOFtY0l3Q14DVZMv77Eo4Nr8O5s+PrF6ZtHhdbzZPnogHp+/vbxYwRl3N8E0z3Nm+8NpS/EqW+O1fcWqb45YI8hVt1UUwEoRELW/KY4DnTxM05xQE4X8J5/ZYk8m/MHmC+w1EJynfUXunAAVRskqthkTmESZDpB0F2QPFd4vhpCfdBpEk86bUE2IX1YhmllavZ8E7oFZPKrE2NOr7Srrmm0gFpbME4glbpIJDU0R8gR+YSHbrjVNSKMlcUP6GHs69VEWcR1B/GHHM+8NegVIrYa7UPXNKx3LIe6WoSyXpcq9sF+o3RSDFcTUJ9CZSleXY7OgPSjaenZqvOnz4XMdReZ+qecqqyx/ljtOs6n+ApDOSurC2uTwJFPiVArr2gqV3tbwX1fVHzfL2Vu1Xag03OpVpC9TXHwaS505VN+A9CzKhlIaYbGkKvyualRYTP3HFvjcaPz38enH0bnV6eauNPtHJo1jvXKv9I4TYRCmWCgWE8RAAEglL1pBiopP5C5ij9qHd8QQ5wbrtTxmVHuwGDlbXdR/mvbFrlZf2w5iSvkqIQ1VwdgCf39+tmUJcGn3GVTyGJhlwdoJ4iKv70g2iBWo/QXJN7SUEUgj7gWQ0ToPFLVkfa+bnU5VHzqFHPKnzEyU0Ytp8phz6F6GNIFouHRhn9IQAHDUsLMLGOQxNyB8x8f9oC+xsYSrHd9viACtqFJop3jMzsIp+ewB49tLmWzfSDXGpXjpeHyqeZkPRgdm383COG8FYZya21mNSf/uKQpNMuv/wU).

# Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [ ] ~I have updated `CHANGES.md` with a short summary of my change~ This will be done in the PR into main
- [x] I have added or updated unit tests to ensure consistent code coverage
- [x] I have update the inline documentation, and included code examples where relevant
- [x] I have performed a self-review of my code

jjhembd commented:
To-do:

- [ ] Fix the spec that is failing on CI and not locally
- [ ] Add `VoxelCellSpec`

ggetz commented:
Thank @jjhembd! The API here is much cleaner :+1:

I saw one issue in the Sandcastle example:

When switching from ellipsoid shape to cylindrical shape, I saw the following output in the console.

```
VoxelPrimitive: Uniform name "u_cylinderUvToRenderBoundsScale" is already defined
VoxelPrimitive: Uniform name "u_cylinderUvToRenderBoundsTranslate" is already defined
VoxelPrimitive: Uniform name "u_cylinderUvToRenderRadiusMin" is already defined
VoxelPrimitive: Uniform name "u_cylinderRenderAngleMinMax" is already defined
VoxelPrimitive: Uniform name "u_cylinderUvToShapeUvRadius" is already defined
VoxelPrimitive: Uniform name "u_cylinderUvToShapeUvHeight" is already defined
VoxelPrimitive: Uniform name "u_cylinderUvToShapeUvAngle" is already defined
VoxelPrimitive: Uniform name "u_cylinderShapeUvAngleMinMax" is already defined
VoxelPrimitive: Uniform name "u_cylinderShapeUvAngleRangeZeroMid" is already defined
```

ggetz commented on packages/engine/Source/Scene/VoxelCell.js:
```diff
@@ -0,0 +1,158 @@
+   * @readonly
+   * @private
+   */
+  metadatas: {
```
A bit of a nitpick, but I think the plural of "metadata" is still "metadata".

ggetz commented on packages/engine/Source/Scene/VoxelCell.js:
```diff
@@ -0,0 +1,158 @@
+});
+
+/**
+ * Returns whether the feature contains this property.
```
```suggestion
 * Returns <code>true</code> if the feature contains this property.
```

jjhembd commented:
Thanks for the feedback @ggetz! I think I addressed it all, and also added the missing specs.

I added an `OrientedBoundingBox` to the `VoxelCell` class, to expose the center point and dimensions of the picked sample.

ggetz commented on packages/engine/Source/Scene/StructuralMetadata.js:
```diff
@@ -5,7 +5,7 @@ import defined from "../Core/defined.js";
  * An object containing structural metadata.
  * <p>
- * See the {@link https://github.com/CesiumGS/glTF/tree/3d-tiles-next/extensions/2.0/Vendor/EXT_structural_metadatas|EXT_structural_metadata Extension} as well as the
+ * See the {@link https://github.com/CesiumGS/glTF/tree/3d-tiles-next/extensions/2.0/Vendor/EXT_structural_metadata|EXT_structural_metadata Extension} as well as the
```
I confirmed this link is now correct. :+1:

ggetz commented on packages/engine/Source/Scene/VoxelCell.js:
```diff
@@ -0,0 +1,268 @@
+) {
+  //>>includeStart('debug', pragmas.debug);
+  Check.typeOf.object("primitive", primitive);
+  Check.typeOf.number("sampleIndex", sampleIndex);
```
Should `tileIndex` be here?

jjhembd replied:
Yes, for consistency. I checked the others because they are needed for subsequent function calls, whereas tileIndex is only reported back to the user. But we might as well check all of them.

ggetz commented on packages/engine/Specs/Scene/PickingSpec.js:
```diff
@@ -250,13 +250,41 @@ describe(
     });
 
+    describe("pickVoxel", function () {
+      it("does not pick undefined window positions", function () {
```
We usually phrase these types of tests as:

```suggestion
      it("throws when window position is undefined", function () {
```

ggetz commented:
Awesome, thanks @jjhembd !
